inst/include/distributions/functors/density_components_base.hpp

```diff
@@ -20,25 +20,84 @@
 
 namespace fims_distributions {
 
+/**
+* Container to hold density components including pointers to density inputs.
+*/
+template <typename Type>
+struct DistributionElementObject {
+  std::string  input_type; /**< string classifies the type of the negative log-likelihood; options are: "priors", "random_effects", and "data" */
+  std::shared_ptr<fims_data_object::DataObject<Type>> observed_values; /**< observed data*/
+  fims::Vector<Type > expected_values; /**< expected value of distribution function*/
+  fims::Vector<Type> *re = NULL; /**< pointer to random effects vector*/
+  fims::Vector<Type> *re_expected_values = NULL; /**< expected value of random effects*/
+  std::vector<fims::Vector<Type>* > priors; /**< vector of pointers where each points to a prior parameter */
+  fims::Vector<Type > x; /**< input value of distribution function for priors or random effects*/
+  //std::shared_ptr<DistributionElementObject<Type>> expected; /**< expected value of distribution function */
+
+    /**
+   * Retrieve element from observed data set, random effect, or prior.
+   * @param i index referencing vector or pointer
+   * @return the reference to the value of the vector or pointer at position i
+   */
+  inline Type get_observed(size_t i) {
+    if(this->input_type == "data"){
+      return observed_values->at(i);
+    }
+    if(this->input_type == "random_effects"){
+      return (*re)[i];
+    }
+    if(this->input_type == "prior"){
+      return (*(priors[i]))[0];
+    }
+    return x[i];
```

Comment:
Is there a way to have an informative abort if the user input something other than the three options?

Reply:
I suggest adding an error to the FIMS log. If FIMS log is set to throw on error, it will abort immediately.

Reply:
I don't like having to require the user to set "throw on error" to make things happen. I would rather the user have the same consistent experience of informative error messages to the console or to a log file without crashing. But, this is a conversation for somewhere outside of this PR as it is a global change that should happen.

Reply:
I think there should also be a dimension check before the vector access.

Reply:
DImension checks are in the density functors (eg. [here](https://github.com/NOAA-FIMS/FIMS/blob/17d4479f812c300044725694a35302393b12c2f2/inst/include/distributions/functors/normal_lpdf.hpp#L48)), but I can see it being more efficient to check dimesions in the get_observed and get_expected functions.

Reply:
That works, but the question is do we just throw invalid_argument, or just log the error, or both?

Reply:
Error early to 🛑 the model if the model won't work.